docs: Module 3 Protobuf (B9lab updates 2-2) (#801)
cosmos/sdk-tutorials: b9lab-content/3-main-concepts/09-protobuf.md

</HighlightBox>

`.proto` files contain data structures called messages. A compiler, `protoc`, interprets the `.proto` file and generates source code in supported languages - C++, C#, Dart, Go, Java, and Python.

### Working with Protocol Buffers

Define a data structure in a `.proto` file. This is a normal text file with descriptive syntax. Data is represented as a message containing name-value pairs called fields.
Then, compile your Protobuf Schema. `.protoc` generates data access classes with accessors for each field in your preferred language according to the command-line options. Accessors include serializing, deserializing, and parsing. 

## Protobuf basics for Go

The [gobs](https://golang.org/pkg/encoding/gob/) package for Go is a comprehensive package for the Go environment, but it doesn’t work well if you need to share information with applications written in other languages. Another challenge is how to contend with fields that may themselves contain information needing to be parsed or encoded.

> For example, a JSON or XML object may contain discrete fields that are stored in a string field. In another example, a time may be stored as two integers representing hours and minutes. A Protobuf encapsulated the necessary conversions in both directions. The generated classes provide getters and setters for the fields and take care of the details for reading and writing the message as a unit.

Reviewer: ```suggestion
For example, a JSON or XML object may contain discrete fields that are stored in a string field. In another example, a time may be stored as two integers representing hours and minutes. A Protobuf encapsulates the necessary conversions in both directions. The generated classes provide getters and setters for the fields and take care of the details for reading and writing the message as a unit.
```

# Protobuf

## Overview

> Protocol buffers, also called Protobuf in the short form, are an open source, extensible,  cross-platform and language-neutral method of serializing object data, primarily for network communication. Libraries for multiple platforms parse a common interface description language to generate source code for encoding and decoding streams of bytes representing structured data. 

Reviewer 2: ```suggestion
Protocol buffers, also called Protobuf in the short form, are an open source, extensible, cross-platform and language-agnostic method of serializing object data, primarily for network communication and storage. Libraries for multiple languages parse a common interface description language to generate source code for encoding and decoding streams of bytes representing structured data. 
```

* List of **Store Keys**: Each module in the Cosmos SDK uses multistore to persist their part of the state. Access to such stores requires a list of keys that are declared in the type definition of the app.
* List of each module’s **Keepers**: A Keeper is an abstract piece in each module to handle the module's interaction with stores, specify references to other modules' keepers, and implement other core functionalities of the module. For cross-module interactions to work, all modules in the Cosmos SDK need to have their keepers declared in the app's type definition and exported as interfaces to other modules so that the keeper's methods of one module can be called and accessed in other modules, when authorized.
* Reference to **codec**: Defaulted to go-amino, the codec in your Cosmos SDK application can be substituted with other suitable encoding frameworks as long as they persist data stores in byte slices and are deterministic.

> Reference to Module Manager: A reference to an object containing a list of the applications modules, known as the Module Manager.

Reviewer 2: ```suggestion
* Reference to the Module Manager: This acts like a router to all the applications modules.
```

Reviewer 3: @xavierlepretre
